tests/rptest/tests/compaction_recovery_test.py

```diff
@@ -79,21 +102,100 @@ def test_index_recovery(self):
-                map(lambda p: len(p.segments) > count and p.recovered(),
-                    partitions))
+class CompactionRecoveryUpgradeTest(RedpandaTest):
+    OLD_VERSION = (22, 2, 8)
 
-        wait_until(lambda: check_partitions(),
-                   timeout_sec=90,
-                   backoff_sec=2,
-                   err_msg="Segments not found")
+    topics = (TopicSpec(partition_count=1,
+                        replication_factor=3,
+                        cleanup_policy=TopicSpec.CLEANUP_COMPACT), )
+
+    def __init__(self, *args, **kwargs):
+        super(CompactionRecoveryUpgradeTest, self).__init__(
+            *args,
+            extra_rp_conf=dict(compacted_log_segment_size=1048576,
+                               log_compaction_interval_ms=1000,
+                               max_compacted_log_segment_size=1,
+                               compaction_ctrl_min_shares=1000,
+                               compaction_ctrl_max_shares=1000),
+            **kwargs)
+
+    def setUp(self):
+        self.installer = self.redpanda._installer
+        self.installer.install(self.redpanda.nodes, self.OLD_VERSION)
+        super(CompactionRecoveryUpgradeTest, self).setUp()
 
-        return partitions
+    @cluster(num_nodes=3, log_allow_list=RESTART_LOG_ALLOW_LIST)
+    def test_index_recovery_after_upgrade(self):
+        """
+        Test compatibility between v1 and v2 compaction index footers.
+        * Redpanda version with v2 footers should be able to recover
+          v1 footers without rebuilding them.
+        * Old redpanda version with v1 footers is not able to recover
+          v2 footers, but at least it should be able to rebuild them.
+        """
+        def get_compacted_segment2mtime(node):
+            partitions = self.redpanda.node_storage(node).partitions(
+                'kafka', self.topic)
+
+            seg2mtime = dict()
+            for p in partitions:
+                for sname, seg in p.segments.items():
+                    if seg.compaction_index:
+                        path = os.path.join(p.path, seg.data_file)
+                        out = p.node.account.ssh_capture(
+                            f"stat --format=%Y {path}")
+                        mtime = ''.join(out).strip()
+                        seg2mtime[path] = int(mtime)
+
+            for k, v in sorted(seg2mtime.items()):
+                self.logger.debug(f"mtime for compacted index {k}: {v}")
+
+            return seg2mtime
+
+        # Restart only 1 node so that cluster feature version doesn't increase
+        # and we can go back to the previous redpanda version.
+        to_restart = self.redpanda.nodes[0]
+        self.logger.info(f"will test node {to_restart.account.hostname}")
+
+        produce_until_segments(self.redpanda, self.topic, 3)
+        seg2mtime_1 = get_compacted_segment2mtime(to_restart)
+        assert len(seg2mtime_1) >= 3
+
+        self.installer.install([to_restart], RedpandaInstaller.HEAD)
```

Comment:
We should upgrade via the intervening versions: jumps straight from 22.2 up to HEAD might not work when head is some future release

Reply:
Yes, there is some missing infrastructure here:
- a helper for stepping through versions in an upgrade
- a definition in the test code of what the intervening versions are between a particular historical version and HEAD (like in https://github.com/redpanda-data/redpanda/pull/8253/files#diff-08ac9c7e3ca9bc4f0dcdb04ab5d5c90bd853ea2175df49ae895d538cff7a3268)

I don't think this has to be a blocker for this PR, but it will need revisiting later to make this test sustainable.